### PR TITLE
Implement support for copy constructors when marshalling in IJW

### LINKED ISSUE #22219: Handling for custom modifiers Microsoft.VisualC.NeedsCopyConstructorModifier and System.Runtime.IsCopyConstructed is not correct
CoreCLR is missing the BlittableValueClassWithCopyCtorMarshaller

The logic in NDirect::MarshalingRequired doesn't detect marshalling needed cases for ByRef parameters. Currently it only implements the logic for Ptr parameters.

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Fixes #22219.

I decided to try simplifying the code when bringing this feature over from .NET Framework. The .NETFX x86 implementation of this marshaler adds a lot of extra code-paths, and intercepts the calling stub to enable it to exactly replicate the behavior of what would be the native code by copy-constructing an object in-place where it goes on the stack for the native call.

Instead of adding all of that extra goo, I decided to keep the implementation much more similar to the non-x86 implementation from .NETFX. Instead of intercepting the call and adding bookkeeping helper stubs, the marshaler just copies to a local in the IL stub, just like non-x86. When calling the native function, it just loads the local onto the IL stack and calls the native function as a normal function. There is a difference there, but I cannot think of a way that the difference is observable to the user.

The non-x86 implementation is identical to the .NETFX implementation.

### CODE REVIEW

File: tests/src/Interop/IJW/CopyConstructorMarshaler/IjwCopyConstructorMarshaler.cpp [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+{
+    int copyCount;
+    int i;
+
+public:
+    A()
+    :copyCount(0)
+    {};
+
+    A(A& other)
```
Comment (AaronRobinsonMSFT) [inline tests/src/Interop/IJW/CopyConstructorMarshaler/IjwCopyConstructorMarshaler.cpp]: Can you also add a test for a copy constructor from a class that has a base class?

File: src/vm/ilmarshalers.cpp [hunk 2]
```diff
@@ -3292,6 +3292,91 @@ ILCriticalHandleMarshaler::ReturnOverride(
...
+
+        // Step 1
+        dwNewValueTypeLocal = pslIL->NewLocal(locDesc);
+
+        // Step 2
+        if (pargs->mm.m_pCopyCtor)
+        {
+            pslIL->EmitLDLOCA(dwNewValueTypeLocal);
+            pslIL->EmitLDARG(argidx);
+            pslIL->EmitCALL(pslIL->GetToken(pargs->mm.m_pCopyCtor), 2, 0);
```
Comment (AaronRobinsonMSFT) [inline src/vm/ilmarshalers.cpp]: Document the 2 and 0 arguments.

File: src/vm/ilmarshalers.cpp [hunk 3]
```diff
@@ -3292,6 +3292,91 @@ ILCriticalHandleMarshaler::ReturnOverride(
...
+
+        // Step 3
+        if (pargs->mm.m_pDtor)
+        {
+            pslIL->EmitLDARG(argidx);
+            pslIL->EmitCALL(pslIL->GetToken(pargs->mm.m_pDtor), 1, 0);
+        }
+#ifdef _TARGET_X86_
+        pslIL->SetStubTargetArgType(&locDesc);              // native type is the value type
+
```
Comment (AaronRobinsonMSFT) [inline src/vm/ilmarshalers.cpp]: Remove line

**Comment by jkoritzinsky** [comment]
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test 
@dotnet-bot test Windows_NT x64 Checked CoreFX Tests
@dotnet-bot test Windows_NT x86 Checked Innerloop Build and Test

**Comment by jkoritzinsky** [comment]
@dotnet-bot test this please

**Comment by jkoritzinsky** [comment]
@dotnet-bot test this please

**Comment by jkoritzinsky** [comment]
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test

**Pull request merged**